Update cp2k recipe to use cmake or the current build system

mtaillefumier opened the pull request:
Offers the possibility to build cp2k with the new cmake build system. commands like this are now supported

spack install cp2k@master build_system=cmake +.....

the recipe supports the following optional functionalities

- superlu, cosma, sirius, spglib, metis, spglib, libxc, libint, cuda/rocm, mkl/openblas/sci (and others), mpi, openmp, dbcsr
- dbcsr is built separately using the currently available recipe.

Two PRs need to be merged to be fully functional (cosma update in spack + one PR in cp2k github).

NB : to be fully functional two PRs need to be merged https://github.com/spack/spack/pull/35615 and one in cp2k repo. It is a temporary limitation at that point but I would like to receive feedback about recipe independently of the two PRs.

limitations are +cosma and +plumed are not properly working: +cosma is fixed (fix by https://github.com/spack/spack/pull/35615), +plumed is a typo in cp2k cmake build system. It is also better to use the cmake build system with the current master as it includes important fix for the cmake build system.

Commit:
Update cp2k recipe to use cmake or the current build system

Offers the possibility to build cp2k with the new cmake build system. commands like this are now supported

spack install cp2k@master build_system=cmake +.....

the recipe supports the following optional functionalities

- superlu, cosma, sirius, spglib, metis, spglib, libxc, libint, cuda/rocm, mkl/openblas/sci (and others), mpi, openmp, dbcsr
- dbcsr is built separately using the currently available recipe.

Two PRs need to be merged to be fully functional (cosma update in spack + one PR in cp2k github).

Commit:
Fix indentation

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -44,15 +50,25 @@ class Cp2k(MakefilePackage, CudaPackage):
         "pexsi",
         default=False,
-        description=("Enable the alternative PEXSI method" "for density matrix evaluation"),
+        description = "Enable the alternative PEXSI method" "for density matrix evaluation",
```
```suggestion
        description = "Enable the alternative PEXSI method for density matrix evaluation",
```

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +79,48 @@ class Cp2k(MakefilePackage, CudaPackage):
         "sirius",
         default=False,
-        description=("Enable planewave electronic structure" " calculations via SIRIUS"),
+        description = "Enable planewave electronic structure" " calculations via SIRIUS",
```
```suggestion
        description = "Enable planewave electronic structure calculations via SIRIUS",
```

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +79,48 @@ class Cp2k(MakefilePackage, CudaPackage):
         "libvori",
         default=False,
-        description=("Enable support for Voronoi integration" " and BQB compression"),
+        description="Enable support for Voronoi integration" " and BQB compression",
```
```suggestion
        description="Enable support for Voronoi integration and BQB compression",
```

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +79,48 @@ class Cp2k(MakefilePackage, CudaPackage):
+        "metis",
+        default=False,
+        description="Enable (par)metis support",
+    )
```
NACK. CP2K only uses (par-)metis via PEXSI, hence a variant does not make sense.

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +79,48 @@ class Cp2k(MakefilePackage, CudaPackage):
+        "superlu-dist",
+        default=False,
+        description=("Enable superlu support"),
     )
```
NACK, same as for metis.

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -153,16 +202,17 @@ class Cp2k(MakefilePackage, CudaPackage):
         depends_on("libxc@5.1.3:5.1", when="@8.2:8")
         depends_on("libxc@5.1.7:5.1", when="@9:2022")
-        depends_on("libxc@6:6.1", when="@2023:")
+        depends_on("libxc@6:6.1", when="@2023:", type=("build", "run"))
```
Why should libxc not get added to the linker?

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -153,16 +202,17 @@ class Cp2k(MakefilePackage, CudaPackage):
 
     with when("+cosma"):
-        depends_on("cosma+scalapack")
+        depends_on("cosma+scalapack", type=("build", "run"))
```
Why should cosma not get added to the linker?

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -220,6 +272,12 @@ class Cp2k(MakefilePackage, CudaPackage):
+        depends_on("dbcsr+openmp", when="+openmp", type=("build", "run"))
+        depends_on("dbcsr+cuda", when="+cuda", type=("build", "run"))
+        depends_on("dbcsr+rocm", when="+rocm", type=("build", "run"))
+
```
same here, why the limitation to build and run but not link?

mtaillefumier replied:
my guess is that I do not really understand what these options are triggering. I will correct all of these.

dev-zero replied:
Do we also have to ensure consistent settings for libxsmm?

Commit:
Fix indentation

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -778,3 +836,106 @@ def check(self):
+                "80": "A100",
+            }
+
+            gpuver = gpu_map[cuda_arch]
```
factor out and share with the makefile one

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -778,3 +836,106 @@ def check(self):
+        if spec["blas"].name in ["intel-mkl", "intel-parallel-studio", "cray-libsci", "openblas"]:
+            if spec["blas"].name in ["intel-mkl", "intel-parallel-studio"]:
+                args += [ "-DCP2K_BLAS_VENDOR=MKL" ]
+                args += [ "-DCP2K_SCALAPACK_VENDOR=MKL" ]
```
Please separate blas and scalapack here, on macOS for example there is no Intel MPI and no ScaLAPACK, hence the combination of MKL for BLAS/LAPACK and netlib-scalapack for scalapack.

dev-zero replied:
In general I wonder whether this kind of explicit spec for the scientific libs is still necessary with CMake or whether Spack provides a restricted enough environment for CMake to auto-discover them correctly?

mtaillefumier replied:
Yes cmake autodetects them automatically. It will take the first one it finds on a system I usually specify things to help cmake finding the dependencies but in the spack case this is known. 

In some case it is necessary to specify these (essl for instance). I have for instance five different implementations of blas / lapack on my system. So specifying one version allows me to test things. 

CP2K cmake build system will still generate a warning if it find mkl but scalapack is set to its default value. So When we know that we use MKL better specify the scalapack version as well. 

macos ? Not sure to understand what's wrong here.

dev-zero replied:
You are assuming here that if MKL is used for BLAS it should also be used for ScaLAPACK.
Which is at least on macOS not possible.

Commit:
Update libvori

Commit:
More typos

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +69,24 @@ class Cp2k(MakefilePackage, CudaPackage):
+        default=False,
+        description="Use SPLA off-loading functionality. Only relevant when CUDA or ROCM are enabled",
+    )
+    variant("pytorch", default=False, description="Enable libtorch support")
+    variant("metis", default=False, description="Enable (par)metis support")
+    variant("quip", default=False, description=("Enable quip support"))
+    variant("superlu-dist", default=False, description=("Enable superlu support"))
```
```suggestion
    variant("pytorch", default=False, description="Enable libtorch support")
    variant("quip", default=False, description=("Enable quip support"))
```

CP2K uses metis and superlu only via PEXSI, variant does not make sense.

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -778,3 +802,106 @@ def check(self):
+            args += ["-DCP2K_USE_LIBXSMM=ON"]
+
+        if spec["blas"].name in ["intel-mkl", "intel-parallel-studio", "cray-libsci", "openblas"]:
+            if spec["blas"].name in ["intel-mkl", "intel-parallel-studio"]:
+                args += ["-DCP2K_BLAS_VENDOR=MKL"]
+                args += ["-DCP2K_SCALAPACK_VENDOR=MKL"]
```
```suggestion
            if spec["blas"].name in ["intel-mkl", "intel-parallel-studio"]:
                args += ["-DCP2K_BLAS_VENDOR=MKL"]
            if spec["scalapack"].name in ["intel-mkl", "intel-parallel-studio"]:
                args += ["-DCP2K_SCALAPACK_VENDOR=MKL"]
```

dev-zero replied:
theoretically, this also applies to all the other cases below

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -778,3 +802,106 @@ def check(self):
+        if "spla" in spec and (spec.satisfies("+cuda") or spec.satisfies("+rocm")):
+            args += ["-DCP2K_USE_SPLA_GEMM_OFFLOADING=ON"]
+
+        if "fftw" in spec:
+            args += ["-DCP2K_USE_FFTW=ON"]
```
```suggestion
        args += ["-DCP2K_USE_FFTW=ON"]
```

I made `fftw-api` a hard requirement for CP2K on Spack. And I guess `USE_FFTW` should be set even if the provider is `libsci` or `mkl`?

RMeli replied:
```suggestion
        args += ["-DCP2K_USE_FFTW3=ON"]
```
I think it should be `FFTW3`

dev-zero commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -778,3 +802,106 @@ def check(self):
+        if "fftw" in spec:
+            args += ["-DCP2K_USE_FFTW=ON"]
+
+        with when("smm=libxsmm"):
+            args += ["-DCP2K_USE_LIBXSMM=ON"]
```
I have to check whether the following works:
```
self.define_from_variant("CP2K_USE_LIBXSMM", "smm=libxsmm"),
```

but in any case we might want to set it to OFF explicitly just to anticipate the CP2K CMake to switch the default at some point.
Also, in DBCSR CMake we have modeled that flag the same way as in in Spack: `-DUSE_SMM=...` to anticipate new/other libsmm providers.

Commit:
Simplify BLAS/LAPACK

Commit:
Simplify BLAS/LAPACK

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -778,3 +826,94 @@ def check(self):
+
+        args += [
+            self.define_from_variant("CP2K_USE_ELPA", "elpa"),
+            self.define_from_variant("CP2K_USE_LIBINT", "libint"),
```
```suggestion
            self.define_from_variant("CP2K_USE_LIBINT2", "libint"),
```

Commit:
Add A100 gpu value

Commit:
Fix typo

Commit:
Add the enable_regtests option

if -DCP2K_ENABLE_REGTESTS=ON (+enable_regtests with spack) then the location of the binary executables will be in the cp2k root directory under exe/build-cmake-*. This option is needed to run the regtests afterwards.

mtaillefumier commented:
A ping to review this PR ;)

alalazo commented:
@dev-zero

Commit:
Minor update

alalazo commented:
@albestro Would you have time to review this PR in the next few days?

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -763,3 +824,104 @@ def check(self):
+        args = []
+
+        if spec.satisfies("+cuda") or spec.satisfies("+rocm"):
+            if spec.satisfies("cuda_target"):
```
```suggestion
            if not spec.satisfies("cuda_arch=none"):
```
?

_But_, it might be clearer to also have separate branches for `spec.satisfies("+cuda")` and `spec.satisfies("+rocm")` (possibly even merging them with the `InstallError` checks above and the `CP2K_USE_ACCEL` branches below?

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -763,3 +824,104 @@ def check(self):
+
+        if spec.satisfies("+cuda") or spec.satisfies("+rocm"):
+            if spec.satisfies("cuda_target"):
+                args += [self.define("CP2K_WITH_GPU", gpu_map[spec.variants["cuda_arch"].value])]
+
+            if spec.satisfies("amdgpu_target"):
+                args += [
+                    self.define("CP2K_WITH_GPU", gpu_map[spec.variants["amdgpu_target"].value])
```
I think `value` is an array with one element at this point:
```suggestion
                args += [self.define("CP2K_WITH_GPU", gpu_map[spec.variants["cuda_arch"].value[0])]

            if spec.satisfies("amdgpu_target"):
                args += [
                    self.define("CP2K_WITH_GPU", gpu_map[spec.variants["amdgpu_target"].value[0])
```

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -763,3 +824,104 @@ def check(self):
+        lapack = spec["lapack"]
+        blas = spec["blas"]
+
+        if spec["blas"].name in ["intel-mkl", "intel-parallel-studio"]:
```
```suggestion
        if blas.name in ["intel-mkl", "intel-parallel-studio"]:
```

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -763,3 +824,104 @@ def check(self):
+        else:
+            args.extend(
+                [
+                    self.define("CP2K_LAPACK_FOUND", "true"),
+                    self.define("CP2K_LAPACK_LINK_LIBRARIES", lapack.libs.joined(";")),
+                    self.define("CP2K_BLAS_FOUND", "true"),
+                    self.define("CP2K_BLAS_LINK_LIBRARIES", blas.libs.joined(";")),
+                    self.define("CP2K_SCALAPACK_FOUND", "true"),
```
```suggestion
                    self.define("CP2K_LAPACK_FOUND", True),
                    self.define("CP2K_LAPACK_LINK_LIBRARIES", lapack.libs.joined(";")),
                    self.define("CP2K_BLAS_FOUND", True),
                    self.define("CP2K_BLAS_LINK_LIBRARIES", blas.libs.joined(";")),
                    self.define("CP2K_SCALAPACK_FOUND", True),
```

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -763,3 +824,104 @@ def check(self):
+        ]
+
+        # we force the use elpa openmp threading support. might need to be revisited though
+        if (spec.satisfies("+openmp") and spec.satisfies("+elpa")) or spec.satisfies(
+            "^elpa+openmp"
+        ):
+            args += ["-DCP2K_ENABLE_ELPA_OPENMP_SUPPORT=ON"]
```
This and the the other args below could be rewritten as
```suggestion
        args += [self.define("CP2K_ENABLE_ELPA_OPENMP_SUPPORT",
                 (spec.satisfies("+openmp") and spec.satisfies("+elpa")) or
                 spec.satisfies("^elpa+openmp"))]
```
but this is a matter of taste.

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -231,14 +266,27 @@ class Cp2k(MakefilePackage, CudaPackage):
+    # rocm already returns an error if amdgpu_target is not defined
 
-    conflicts("+cuda", when="cuda_arch=none", msg=cuda_msg)
+    with when("+cuda"):
+        for arch in CudaPackage.cuda_arch_values:
+            if arch not in supported_cuda_arch_list:
+                conflicts("+cuda", when="cuda_arch={0}".format(arch), msg=cuda_msg)
```
Do you want to add similar conflicts for the unsupported rocm architectures?

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -575,12 +623,25 @@ def edit(self, spec, prefix):
-            cuda_arch = spec.variants["cuda_arch"].value[0]
-            if cuda_arch:
-                gpuver = {"35": "K40", "37": "K80", "60": "P100", "70": "V100"}[cuda_arch]
+            gpuver = gpu_map[spec.variants["cuda_arch"].value]
+            if cuda_arch == "35" and spec.satisfies("+cuda_arch_35_k20x"):
```
```suggestion
            cuda_arch = spec.variants["cuda_arch"].value[0]
            gpuver = gpu_map[cuda_arch]
            if cuda_arch == "35" and spec.satisfies("+cuda_arch_35_k20x"):
```
?

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -575,12 +623,25 @@ def edit(self, spec, prefix):
+            acc_flags_var = "NVFLAGS"
+            cppflags += ["-D__ACC"]
+            cppflags += ["-D__DBCSR_ACC"]
+            gpuver = gpu_map[spec.variants["amdgpu_target"].value]
```
```suggestion
            gpuver = gpu_map[spec.variants["amdgpu_target"].value[0]]
```

msimberg replied:
Also, does `gpuver` need to be used somewhere in the `+rocm` case? I can't see it being used but I may have missed it...

Commit:
more fixes

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Mikael Simberg <mikael.simberg@iki.fi>

Commit:
small changes

msimberg commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -231,14 +267,46 @@ class Cp2k(MakefilePackage, CudaPackage):
-            conflicts("+cuda", when="cuda_arch={0}".format(arch), msg=cuda_msg)
+    supported_cuda_arch_list = ("35", "37", "60", "70", "80")
+    supported_rocm_arch_list = (
+        "60",
+        "70",
+        "80",
```
I guess these don't belong in the rocm list?

msimberg reviewed (changes requested):
One more nit from me, looks good to me otherwise.

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +73,31 @@ class Cp2k(MakefilePackage, CudaPackage):
+        " relevant when CUDA or ROCM are enabled",
+    )
+    variant("pytorch", default=False, description="Enable libtorch support")
+    variant("quip", default=False, description=("Enable quip support"))
```
```suggestion
    variant("quip", default=False, description="Enable quip support")
```

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +73,31 @@ class Cp2k(MakefilePackage, CudaPackage):
+        "spla",
+        default=False,
+        description="Use SPLA off-loading functionality. Only"
+        " relevant when CUDA or ROCM are enabled",
```
```suggestion
        description="Use SPLA off-loading functionality. Only relevant when CUDA or ROCM are enabled",
```

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +73,31 @@ class Cp2k(MakefilePackage, CudaPackage):
+    variant(
+        "enable_regtests",
+        default=False,
+        description="Configure cp2k to run the regtests afterwards."
+        " It build cp2k normally but put the executables in exe/cmake-build-* instead of the"
+        " conventional location. This option is only relevant when regtests need to be run.",
```
Same "style fix" about split string here?

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -231,14 +267,46 @@ class Cp2k(MakefilePackage, CudaPackage):
+        "gfx90a:xnack+": "Mi250",
+    }
+    cuda_msg = "cp2k only supports cuda_arch {0}".format(supported_cuda_arch_list)
+    rocm_msg = "cp2k only supports andgpu_target {0}".format(supported_rocm_arch_list)
```
```suggestion
    rocm_msg = "cp2k only supports amdgpu_target {0}".format(supported_rocm_arch_list)
```

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -5,23 +5,33 @@
 import os.path
+import sys
 
+import spack.platforms
 import spack.util.environment
+import spack.util.executable
+from spack.build_environment import dso_suffix
```
What are these needed for?

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -576,11 +644,25 @@ def edit(self, spec, prefix):
 
-                if cuda_arch == "35" and spec.satisfies("+cuda_arch_35_k20x"):
-                    gpuver = "K20X"
+        if spec.satisfies(+rocm) and spec.satisfies("@2022:"):
```
FIY, I think this is equivalent to something like

```suggestion
        if "@2022: +rocm" in spec:
```

Commit:
Remove any reference to nvidia architecture in the rocm list

mtaillefumier committed:
Merge branch 'develop' into cp2k_update_2

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -763,3 +845,114 @@ def check(self):
+
+        if "+rocm" in spec:
+            if len(spec.variants["amdgpu_target"].value) > 1:
+                raise InstallError("CP2K supports only one amdgpu_target at a time")
```
```suggestion
                raise InstallError("CP2K supports only one amdgpu_target at a time.")
```

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -763,3 +845,114 @@ def check(self):
+            self.define(
+                "CP2K_ENABLE_ELPA_OPENMP_SUPPORT",
+                (spec.satisfies("+openmp") and spec.satisfies("+elpa"))
+                or spec.satisfies("^elpa+openmp"),
```
This is just an example where readability might improve using the `in` operator
```suggestion
                ("+elpa +openmp" in spec) or ("^elpa +openmp" in spec),
```

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -147,13 +173,13 @@ class Cp2k(MakefilePackage, CudaPackage):
-        depends_on("pkgconfig", type="build", when="@7.0:")
-        depends_on("libxc@2.2.2:3", when="@:5", type="build")
-        depends_on("libxc@4.0.3:4", when="@6.0:6.9", type="build")
+        depends_on("pkgconfig", when="@7.0:")
```
not sure about the others (I see that a discussion on this topic has already occurred with @dev-zero), but probably `pkgconfig` is actually needed just at build time...isn't it?

albestro commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -576,11 +644,25 @@ def edit(self, spec, prefix):
-                    gpuver = "K20X"
+        if spec.satisfies(+rocm) and spec.satisfies("@2022:"):
+            libs += [
+                "-L{}".format(spec["rocm"].libs.directories[0]),
+                "-L{}/stubs".format(spec["rocm"].libs.directories[0]),
+                "-lhipblas",
+                "-lhipfft",
+                "-lstdc++",
```
I don't know the reason why you need to explicitly instruct the rocm dependency like this but FIY

```bash
>>> from spack.spec import Spec
>>> s = Spec("hwloc") # just a random package with a dependency
>>> s.concretize()
>>> dep = s["ncurses"]
>>> dep.libs.search_flags
'-L/Users/ialberto/spack/opt/spack/darwin-ventura-m2/clang-16.0.3/ncurses-6.4-pylthzrstz72hhppg4oiyr3glvziyyur/lib'
>>> dep.libs.link_flags
'-lncurses -ltinfo -lncursesw -ltinfow'
>>> dep.libs.ld_flags
'-L/Users/ialberto/spack/opt/spack/darwin-ventura-m2/clang-16.0.3/ncurses-6.4-pylthzrstz72hhppg4oiyr3glvziyyur/lib -lncurses -ltinfo -lncursesw -ltinfow'
```

albestro reviewed (commented):
Super work @mtaillefumier!

I don't have knowledge about CP2K, so from my side just minor comments and suggestions on style/available helpers.

About `"+variant in spec"` vs `spec.satisfies("+variant")` styles, I don't think there is any preferred method so no change is required, but in general I would try to stick to one (and personally I prefer the former one, more coincise and readable most of the times in my opinion).

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

mtaillefumier committed:
Update var/spack/repos/builtin/packages/cp2k/package.py

Co-authored-by: Alberto Invernizzi <9337627+albestro@users.noreply.github.com>

Commit:
Final reformating

Commit:
Update py-fypp

RMeli commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -763,3 +836,113 @@ def check(self):
+        spec = self.spec
+        args = []
+
+        gpu_map = {
+            "35": "K40",
+            "37": "K80",
+            "60": "P100",
+            "70": "V100",
+            "80": "A100",
+            "gfx906": "Mi50",
+            "gfx908": "Mi100",
+            "gfx90a": "Mi250",
+            "gfx90a:xnack-": "Mi250",
+            "gfx90a:xnack+": "Mi250",
+        }
```
The same `gpu_map` is used above, is it worth defining it in one place (globally)?

lorisercole commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -198,27 +225,36 @@ class Cp2k(MakefilePackage, CudaPackage):
-    depends_on("python@3.6:", when="@7:+cuda", type="build")
+    depends_on("py-numpy", when="@7:+cuda")
+    depends_on("python@3.6:", when="@7:+cuda")
+    depends_on("py-fypp")
```
I just noticed that the `py-fypp` dependency causes `cp2k` to depend on `python`. Is that really necessary?
Or is it just a build dependency? @mtaillefumier

RMeli replied:
`py-fypp` should be a build dependency AFAIK.

giordano commented on var/spack/repos/builtin/packages/cp2k/package.py:
```diff
@@ -63,15 +73,31 @@ class Cp2k(MakefilePackage, CudaPackage):
+        description="Use SPLA off-loading functionality. Only relevant when CUDA or ROCM"
+        " are enabled",
+    )
+    variant("pytorch", default=False, description="Enable libtorch support")
```
When `pytorch` is enabled, there is no extra dependency on `py-torch`.  How is this supposed to work?  Was it an oversight?

RMeli replied:
@giordano this looks like an oversight indeed.
